qiskit_optimization/algorithms/qrao/magic_rounding.py

```diff
@@ -0,0 +1,448 @@
+        # probability of picking the corresponding magic basis on that qubit.
+        basis_probs = []
+        for dvars in q2vars:
+            if vars_per_qubit == 3:
+                x = 0.5 * (1 - clipped_expectation_values[dvars[0]])
+                y = 0.5 * (1 - clipped_expectation_values[dvars[1]]) if (len(dvars) > 1) else 0
+                z = 0.5 * (1 - clipped_expectation_values[dvars[2]]) if (len(dvars) > 2) else 0
+                # ppp:   mu±   = .5(I ± 1/sqrt(3)( X + Y + Z))
+                # pmm: X mu± X = .5(I ± 1/sqrt(3)( X - Y - Z))
+                # mpm: Y mu± Y = .5(I ± 1/sqrt(3)(-X + Y - Z))
+                # mmp: Z mu± Z = .5(I ± 1/sqrt(3)(-X - Y + Z))
+                # fmt: off
+                ppp_mmm =   x   *   y   *   z   + (1-x) * (1-y) * (1-z)
+                pmm_mpp =   x   * (1-y) * (1-z) + (1-x) *   y   *   z
+                mpm_pmp = (1-x) *   y   * (1-z) +   x   * (1-y) *   z
+                ppm_mmp =   x   *   y   * (1-z) + (1-x) * (1-y) *   z
+                # fmt: on
+                basis_probs.append([ppp_mmm, pmm_mpp, mpm_pmp, ppm_mmp])
+            elif vars_per_qubit == 2:
+                x = 0.5 * (1 - clipped_expectation_values[dvars[0]])
+                z = 0.5 * (1 - clipped_expectation_values[dvars[1]]) if (len(dvars) > 1) else 0
+                # pp:   xi±   = .5(I ± 1/sqrt(2)( X + Z ))
+                # pm: X xi± X = .5(I ± 1/sqrt(2)( X - Z ))
+                # fmt: off
+                pp_mm =   x   *   z   + (1-x) * (1-z)
+                pm_mp =   x   * (1-z) + (1-x) *   z
+                # fmt: on
+                basis_probs.append([pp_mm, pm_mp])
+            elif vars_per_qubit == 1:
+                basis_probs.append([1.0])
+        bases_ = [
+            [
+                self.rng.choice(2 ** (vars_per_qubit - 1), p=[p.real for p in probs])
+                for probs in basis_probs
+            ]
+            for _ in range(self._shots)
+        ]
+        bases, basis_shots = np.unique(bases_, axis=0, return_counts=True)
+        return bases, basis_shots
+
+    def round(self, rounding_context: RoundingContext) -> MagicRoundingResult:
+        """Perform magic rounding using the given RoundingContext.
+
+        Args:
+            rounding_context: The context containing the information needed for the rounding.
+
+        Returns:
+            MagicRoundingResult: The results of the magic rounding process.
+
+        Raises:
+            NotImplementedError: If the circuit is not available for magic rounding.
+            ValueError: If the sampler is not configured with a number of shots.
+        """
+        expectation_values = rounding_context.expectation_values
+        circuit = rounding_context.circuit
+        q2vars = rounding_context.encoding.q2vars
+        var2op = rounding_context.encoding.var2op
+        vars_per_qubit = rounding_context.encoding.max_vars_per_qubit
+
+        if circuit is None:
+            raise NotImplementedError(
```

Comment:
I commented before on a NotImplementedError - similar thing here. Interestingly here just below it does a similar sort of check on shots but uses ValueError

Reply:
you're right; a `RuntimeError` or similar is more appropriate here.